CMakeLists.txt

```diff
@@ -95,162 +142,154 @@ macro(clangformat_targets)
-      		  # Get better stack traces in CI and Debug.
-      		  target_link_libraries(snmalloc_lib INTERFACE "-rdynamic")
-            add_compile_options(-g)
-    	    endif()
-    endif()
-
-    if(SNMALLOC_OPTIMISE_FOR_CURRENT_MACHINE)
-      check_cxx_compiler_flag(-march=native SUPPORT_MARCH_NATIVE)
-      if (SUPPORT_MARCH_NATIVE)
-        add_compile_options(-march=native)
-      else()
-        message(WARNING "Compiler does not support `-march=native` required by SNMALLOC_OPTIMISE_FOR_CURRENT_MACHINE")
-      endif()
-    endif()
-
-    find_package(Backtrace)
-    if(${Backtrace_FOUND})
-      target_compile_definitions(snmalloc_lib INTERFACE -DBACKTRACE_HEADER="${Backtrace_HEADER}")
-      target_link_libraries(snmalloc_lib INTERFACE ${Backtrace_LIBRARIES})
-      target_include_directories(snmalloc_lib INTERFACE ${Backtrace_INCLUDE_DIRS})
-    endif()
+# To build with just the header library target define SNMALLOC_HEADER_ONLY_LIBRARY
+if(NOT SNMALLOC_HEADER_ONLY_LIBRARY)
 
-  endif()
-
-  macro(subdirlist result curdir)
+  function(subdirlist result curdir)
     file(GLOB children CONFIGURE_DEPENDS LIST_DIRECTORIES true RELATIVE ${curdir} ${curdir}/* )
     set(dirlist "")
     foreach(child ${children})
       if(IS_DIRECTORY ${curdir}/${child})
         list(APPEND dirlist ${child})
       endif()
     endforeach()
-    set(${result} ${dirlist})
-  endmacro()
+    set(${result} ${dirlist} PARENT_SCOPE)
+  endfunction()
 
-  if(CMAKE_VERSION VERSION_LESS 3.14)
-    set(CMAKE_REQUIRED_LIBRARIES -fuse-ld=lld)
-  else()
-    set(CMAKE_REQUIRED_LINK_OPTIONS -fuse-ld=lld)
-  endif()
+  set(CMAKE_REQUIRED_LINK_OPTIONS -fuse-ld=lld)
   check_cxx_source_compiles("int main() { return 1; }" LLD_WORKS)
+  if (LLD_WORKS)
+    message(STATUS "Using LLD to link snmalloc shims")
+  endif()
 
-  macro(add_shim name type)
+  function(add_shim name type)
     add_library(${name} ${type} ${ARGN})
-    target_link_libraries(${name} snmalloc_lib)
-    if(NOT MSVC)
-      target_compile_definitions(${name} PRIVATE "SNMALLOC_EXPORT=__attribute__((visibility(\"default\")))")
-    endif()
+    target_link_libraries(${name} snmalloc)
     set_target_properties(${name} PROPERTIES CXX_VISIBILITY_PRESET hidden)
+    target_compile_definitions(${name} PRIVATE "SNMALLOC_USE_${SNMALLOC_CLEANUP}")
 
-    # Ensure that we do not link against C++ stdlib when compiling shims.
-    if(NOT MSVC)
-      set_target_properties(${name} PROPERTIES LINKER_LANGUAGE C)
-      if (LLD_WORKS)
-        message("Using LLD.")
+    if(MSVC)
+      target_compile_options(${name} PRIVATE /Zi /W4 /WX /wd4127 /wd4324 /wd4201)
+      set(CMAKE_EXE_LINKER_FLAGS_RELEASE "${CMAKE_EXE_LINKER_FLAGS_RELEASE} /DEBUG")
+    else()
+      target_compile_definitions(${name} PRIVATE "SNMALLOC_EXPORT=__attribute__((visibility(\"default\")))")
+      target_compile_options(${name} PRIVATE
+        -fno-exceptions -fno-rtti -fomit-frame-pointer -ffunction-sections
+        -Wall -Wextra -Werror -Wundef
```

Comment:
Shouldn't the tests also be built with warnings?

Reply:
Yes.  Fixed.